src/dictionary.c

```diff
@@ -369,6 +369,19 @@ SEXP vec_match_params(SEXP needles,
                                 DF_FALLBACK_quiet,
                                 &_);
   PROTECT_N(type, &nprot);
+  // can exit early here. Following operations cannot fail:
+  // * `vec_cast()` must work
```

Comment:
I think there is a chance that `vec_cast()` might fail, e.g. in case of lossy coercions.

Reply:
I thought `vec_ptype2()` returns the richer type so that we can safely cast. In particular I would not expect to get a lossy coercion there. Is this not the case?

Reply:
I think you're right in general, at least for lossy coercions. But in principle there might be types where casting to the common type might fail. E.g. a strict factor type that where `ptype(strict_factor, character) == strict_factor`, and then `vec_cast()` might fail.

That said, it might be a good reason to decide that strict types shouldn't be coercible this way, and either fail (making it less practical for users) or lose their strictness (but that's counter to the idea of a safe strict type).